File: listacircular/listacircular.cs
using System;

class Nodo {
    public int dato;
    public Nodo sig;

    public Nodo(int d){
        dato = d;
        sig = null;
    }
}

class ListaCircular
{

    Nodo cabeza = null;

    public void InsertarInicio(int dato)
    {
        Nodo nuevo = new Nodo(dato);

        if (cabeza == null)
        {
            nuevo.sig = nuevo;
            cabeza = nuevo;
            return;
        }

        Nodo temp = cabeza;
        while (temp.sig != cabeza) temp = temp.sig;

        nuevo.sig = cabeza;
        temp.sig = nuevo;
        cabeza = nuevo;
    }

    public void InsertarFinal(int dato)
    {
        Nodo nuevo = new Nodo(dato);

        if (cabeza == null)
        {
            nuevo.sig = nuevo;
            cabeza = nuevo;
            return;
        }

        Nodo temp = cabeza;
        while (temp.sig != cabeza) temp = temp.sig;

        temp.sig = nuevo;
        nuevo.sig = cabeza;
    }

    public void InsertarPosicion(int dato, int pos)
    {
        if (pos == 0)
        {
            InsertarInicio(dato);
            return;
        }

        Nodo nuevo = new Nodo(dato);
        Nodo temp = cabeza;

        for (int i = 1; i < pos && temp.sig != cabeza; i++)
            temp = temp.sig;

        nuevo.sig = temp.sig;
        temp.sig = nuevo;
    }

    public void EliminarInicio()
    {
        if (cabeza == null) return;

        if (cabeza.sig == cabeza)
        {
            cabeza = null;
            return;
        }

        Nodo fin = cabeza;
        while (fin.sig != cabeza) fin = fin.sig;

        cabeza = cabeza.sig;
        fin.sig = cabeza;
    }

    public void EliminarFinal()
    {
        if (cabeza == null) return;

        if (cabeza.sig == cabeza)
        {
            cabeza = null;
            return;
        }

        Nodo temp = cabeza;
        Nodo ant = null;

        while (temp.sig != cabeza)
        {
            ant = temp;
            temp = temp.sig;
        }

        ant.sig = cabeza;
    }

    public void EliminarPosicion(int pos)
    {
        if (cabeza == null) return;

        if (pos == 0)
        {
            EliminarInicio();
            return;
        }

        Nodo temp = cabeza;
        Nodo ant = null;

        for (int i = 0; i < pos && temp.sig != cabeza; i++)
        {
            ant = temp;
            temp = temp.sig;
        }

        ant.sig = temp.sig;
    }

    public void Mostrar()
    {
        if (cabeza == null) return;

        Nodo temp = cabeza;
        do
        {
            Console.Write(temp.dato + " -> ");
            temp = temp.sig;
        } while (temp != cabeza);

        Console.WriteLine("(regresa)");
    }
    public static void Main(string[] args)
    {
        seguir = true;
        
        do
        {
            ListaCircular lista = new ListaCircular();
            Console.WriteLine("Lista Circular:");
            Console.WriteLine("Elija una opcion:");
            Console.WriteLine("1. Insertar al inicio");
            Console.WriteLine("2. Insertar al final");
            Console.WriteLine("3. Insertar en posicion");
            Console.WriteLine("4. Eliminar al inicio");
            Console.WriteLine("5. Eliminar al final");
            Console.WriteLine("6. Eliminar en posicion");
            Console.WriteLine("7. Mostrar lista");
            Console.WriteLine("8. Salir");
            int opcion = Convert.ToInt32(Console.ReadLine());
            switch (opcion)
            {
                case 1:
                    insertarInicio();
                    break;
                case 2:
                    insertarFinal();
                    break;
                case 3:
                Console.WriteLine("Ingrese el dato a insertar:");
                    int dato = Convert.ToInt32(Console.ReadLine());
                    Console.WriteLine("Ingrese la posicion:");
                    int pos = Convert.ToInt32(Console.ReadLine());
                    insertarPosicion(dato, pos);
                    break;
                case 4:
                    eliminarInicio();
                    break;
                case 5:
                    eliminarFinal();
                    break;
                case 6:
                    Console.WriteLine("Ingrese la posicion a eliminar:");
                    int posicion = Convert.ToInt32(Console.ReadLine());
                    eliminarPosicion(posicion); 
                    break;
                case 7:
                    mostrarLista();
                    break;
                case 8:
                    seguir = false;
                    break;
                default:
                    Console.WriteLine("Opcion no valida");
                    break;
            }
        } while (seguir);

        
    }
}

File: listacircular/listacircular.py
class Nodo:
    def __init__(self, dato):
        self.dato = dato
        self.sig = None

class ListaCircular:

    def __init__(self):
        self.cabeza = None

    def insertar_inicio(self, dato):
        nuevo = Nodo(dato)

        if self.cabeza is None:
            nuevo.sig = nuevo
            self.cabeza = nuevo
            return

        temp = self.cabeza
        while temp.sig != self.cabeza:
            temp = temp.sig

        nuevo.sig = self.cabeza
        temp.sig = nuevo
        self.cabeza = nuevo

    def insertar_final(self, dato):
        nuevo = Nodo(dato)

        if self.cabeza is None:
            nuevo.sig = nuevo
            self.cabeza = nuevo
            return

        temp = self.cabeza
        while temp.sig != self.cabeza:
            temp = temp.sig

        temp.sig = nuevo
        nuevo.sig = self.cabeza

    def insertar_posicion(self, dato, pos):
        if pos == 0:
            self.insertar_inicio(dato)
            return

        nuevo = Nodo(dato)
        temp = self.cabeza

        for _ in range(1, pos):
            if temp.sig == self.cabeza:
                break
            temp = temp.sig

        nuevo.sig = temp.sig
        temp.sig = nuevo

    def eliminar_inicio(self):
        if self.cabeza is None:
            return

        temp = self.cabeza

        if temp.sig == temp:
            self.cabeza = None
            return

        fin = self.cabeza
        while fin.sig != self.cabeza:
            fin = fin.sig

        self.cabeza = self.cabeza.sig
        fin.sig = self.cabeza

    def eliminar_final(self):
        if self.cabeza is None:
            return

        temp = self.cabeza

        if temp.sig == temp:
            self.cabeza = None
            return

        ant = None
        while temp.sig != self.cabeza:
            ant = temp
            temp = temp.sig

        ant.sig = self.cabeza

    def eliminar_posicion(self, pos):
        if self.cabeza is None:
            return

        if pos == 0:
            self.eliminar_inicio()
            return

        temp = self.cabeza
        ant = None

        for _ in range(pos):
            ant = temp
            temp = temp.sig
            if temp == self.cabeza:
                break

        ant.sig = temp.sig

    def mostrar(self):
        if self.cabeza is None:
            return

        temp = self.cabeza
        while True:
            print(temp.dato, "->", end=" ")
            temp = temp.sig
            if temp == self.cabeza:
                break

        print("(regresa)")
seguir = True
while seguir:
    lista = ListaCircular()
    print("Lista Circular:")
    print("Elija una opcion:")
    print("1. Insertar al inicio")
    print("2. Insertar al final")
    print("3. Insertar en posicion")
    print("4. Eliminar al inicio")
    print("5. Eliminar al final")
    print("6. Eliminar en posicion")
    print("7. Mostrar lista")
    print("8. Salir")

    opcion = int(input())

    if opcion == 1:
        dato = int(input("Ingrese el dato a insertar al inicio: "))
        lista.insertar_inicio(dato)
    elif opcion == 2:
        dato = int(input("Ingrese el dato a insertar al final: "))
        lista.insertar_final(dato)
    elif opcion == 3:
        dato = int(input("Ingrese el dato a insertar: "))
        pos = int(input("Ingrese la posicion: "))
        lista.insertar_posicion(dato, pos)
    elif opcion == 4:
        lista.eliminar_inicio()
    elif opcion == 5:
        lista.eliminar_final()
    elif opcion == 6:
        pos = int(input("Ingrese la posicion a eliminar: "))
        lista.eliminar_posicion(pos)
    elif opcion == 7:
        lista.mostrar()
    elif opcion == 8:
        seguir = False
    else:
        print("Opcion no valida")